dev/com.ibm.websphere.appserver.features/test/src/com/ibm/ws/feature/tests/VisibilityTest.java

```diff
@@ -273,14 +273,88 @@ public void testVisibility2() {
             File featureFile = featureInfo.getFeatureFile();
             String fileName = featureFile.getAbsolutePath();
             fileName = fileName.replace('\\', '/');
-            if (!fileName.contains("/visibility/" + visibility + "/")) {
+            String expectedPathName = "/visibility/" + visibility + "/";
+            if ("public".equals(visibility)) {
+                expectedPathName += (featureInfo.getShortName() + '/');
+            }
+            expectedPathName += (featureName + ".feature");
+            if (!fileName.endsWith(expectedPathName)) {
+                errorMessage.append("Found issues with " + featureName + '\n');
+                errorMessage.append("     Feature is not in the expected directory " + expectedPathName + ".\n");
+                errorMessage.append("     The feature's visibility " + visibility + " may not match the directory.\n");
+                if ("public".equals(visibility)) {
+                    errorMessage.append("     AND/OR the feature short name " + featureInfo.getShortName() + " may not match the directory.\n");
+                }
+            }
+        }
+        if (errorMessage.length() != 0) {
+            Assert.fail("Found features that appear to be in the wrong directory based off of their settings: " + '\n' + errorMessage.toString());
+        }
+    }
+
+    /**
+     * Tests that an auto feature has more than one feature in its filter.
+     * Tests that an auto feature doesn't depend on another auto feature.
+     */
+    @Test
+    public void testAutoFeatures() {
+        StringBuilder errorMessage = new StringBuilder();
+        for (Entry<String, FeatureInfo> entry : features.entrySet()) {
+            String featureName = entry.getKey();
+            FeatureInfo featureInfo = entry.getValue();
+
+            if (!featureInfo.isAutoFeature()) {
+                continue;
+            }
+
+            String[] filterFeatures = featureInfo.getAutoFeatures();
+            if (filterFeatures.length == 1) {
```

Comment:
I think it is theoretically possible for there to be no auto-features. So should this check be `<=1`?

Reply:
I have added the <= 1 logic